NEP 42: Technical decisions for new DTypes

seberg opened the pull request:
[Rendered version as of 2020-06-08](https://github.com/numpy/numpy/blob/20ccf8f2cafc9efdcf8146332998745ffbdfd53d/doc/neps/nep-0042-new-dtypes.rst)

This NEP may require some thought and technical changes. There are also a few other things as to how we exactly implement e.g. lookup of which cast to use.

I currently envision that we accept NEP 41, with the decision to start on this NEP. However, we keep most of the proposal here private API, and thus somewhat flexible to changes before we finalize the API.
Some changes (i.e. the exact "methods", we need to define) should probably not be included here, but rather incrementally added later. Maybe by modifying the document, or adding a new NEP depending on the size of the change.

seberg committed:
NEP: Technical decisions for new DTypes

mhvk reviewed (commented):
@seberg - this is all super-interesting so I had a good read.  Which mostly was enlightening though as always one gets questions when reading about the implementation!  

I guess my main suggestion is to work out how `Datetime` would work in more detail. In particular, do include it in the figure, to show what it inherits from and where exactly the unit lives. I ask since for me at least it is not yet clear how this would work. (Here, do include `Datetime` and `timedelta` - conceptually, they are quite different.)

More generally, despite the very logical argument that a unit should live on the dtype and not the array, it does seem that there still is a problem in that one is now combining the concern of how bytes are organized in memory into a number with the concern of what that number actually means.

I guess more concretely, one can (and in astropy we do) have a float with units of nanoseconds; this suggests that datetime does not specifically inherit from integer, but is a combination of a unit class and an integer type.

Separately, it would be nice to solve what `longdouble` and `float128` actually mean!

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+explicitly by the user.
+This removes the need to trust downstream packages to take care not to have
+the potential for modifying existing code.
+**TODO: Maybe I actually like that more right now? Removes some convenience in (rare?) cases, but also removes the annoying global state...**
```
Woops, forgot to encorporate that one already :). There is another interesting thing, which is that the actual cast (when you get the cast function) is just a view (basically `PyArray_EquivalentTypes`. I think this is better handled, by signalling that a view is OK when we get the actual casting function. Have to think about how to signal that however.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+   parsing logic.   
+3. ``__discover_descr_from_pyobject__(cls, obj) -> dtype``: A classmethod that
+   returns the correct descriptor given the input object.
+   *Note that only flexible DTypes have to implement this*.
```
Yes, there would be a default which just does `type.__new__()`, or maybe a specific `type.__default_descriptor__()`.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+``knwon_python_types[type] = DType``.
+NumPy currently uses a small hard-coded a mapping and conversion of numpy scalars
+(inheriting from ``np.generic``) to achieve this, however, this forces a certain
+structure on the associated ``type``.
```
Right, Python integers would probably map to -> `AbstractPyInt[value]` (a class), this can then return an actual dtype when asked for it (i.e. integer). At the same time, `common_dtype` can be defined correctly for it *and* whatever machinery we use for UFuncs has the information available needed for value based casting. Partially, I thought I can defer it to the UFuncs, but for array coercion it is also important.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+casting currently implemented for scalars, giving even the possibility of
+expanding it to e.g. list inputs ``np.add(np.array([8], dtype="uint8"), [4])``
+giving a ``uint8`` result.
+This is mainly related to the question to how the common dtype is found above.
```
Right, class hierarchy (or at least having the classes to begin with) is the first step. I would like to push that forward already before settling on this, by agreeing on NEP 41. May have to move some of the information there though...

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+   of the same DType are given. For builtin dtypes (that are not flexible), this
+   currently always returns ``self`` (but ensures native byte order).
+   This is to preserve metadata. We can thus provide a default implementation
+   for non-flexible user dtypes.
```
`__len__` is only generally defined for a DType instance! Abstract dtypes cannot be instantiated. Of course a `Float128` could inform you that all instances are 128bit, but there is no reason it has to.

I am not sure about `dtype1 & dtype2`. It might make sense to note such possibilities at some point though, if only to discourage downstream to implement dtypes which use such API for something else.

seberg replied:
Did not change anything here.
About the `__call__` I think it would be OK for an abstract DType to actually return a concrete DType instance which it considers a subclass of itself (i.e. similar to a class factory). Creating the scalar should rather be through `DType.type()` or another classmethod I think.

The reuse of existing slots is interesting... We could just do that, or at least document that we may want to do that at some point ("please do not fill them in", we could even enforce that most operators are not defined by users, e.g. we have authority over what `==` does).

The last part should maybe be added, even if it is a detail?

mhvk replied:
Interesting ideas - I think it would be good to add that we are still considering to what extent existing slots can be useful.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+However, the issue with this approach is that a newly added DType can change
+the behaviour of an existing program. For example, a new ``int24`` would be
+the first valid common type for ``int16`` and ``uint16``, demoting the currently
+defined behaviour of ``int32``.
```
My current sense is: Lets not go there :). These type of things run into scoping issues quickly. So if a user wants `uint16 + int16` to be `int24` they probably should be either using their own `uint16`, or use `dtype=int24`.
I am also OK with adding it at some point, but I am scared of it, and I would prefer if we can agree to *not* do anything in this direction right now.

mhvk replied:
OK, makes sense not to start with this!

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+pass around a specialized casting function directly if so wished.
+
+In the future, we may considering adding a way to spell out that specific
+casts are known to be *not* possible.
```
Currently, I am thinking: Make it analogous to ufuncs, *except* for the loop discovery itself. You could just have a `np.cast` ufunc as well. But I somewhat tend to casting being special enough to not do that. Plus, it ensures that (whatever you do for ufuncs) casting is a bit more contraint.

I probably have to add some more reasons for that decision though.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+      int abstract;             /* Is the dtype abstract? */
+      int flags;                /* Currently only the "needs API" flag */
+      /* NULL terminated CastingImpl; is copied and references are stolen */
+      CastingImpl *castingimpls[];
```
You are not inheriting from anything except DType here (unless its abstract). So yes, I could make an `AbstractDType` as a subclass (or superclass) of `np.dtype`, which is used to indicate that it is abstract.
My current thinking is that I want to be able to ensure that any abstract dtype cannot be simply instantiated, and this might be easiest. But hmmmm. The `flexible` part is unnecessary, but should allow some optimizations and also sanity checks because "flexible" dtypes require more logic.

seberg replied:
We can use ABCs, I am not sure it is valuable though (plus I am a bit scared of python changing how they define ABCs). In fact ABCs are *effectively* defined in Python right now (they are optimized using C, but they do not use an `ExtenionMetaClass` as I do (that does not make things harder maybe though).

On the other hand, I expect the actual need we have is only overriding `isinstance`, etc. so I am not sure it is valuable to use ABCs, it seems like it may add more complexity then it solves.

mhvk replied:
Indeed, not sure it adds enough, but really it was meant as encouragement (for you, I fear...) to look what they actually do so that we do not reinvent the wheel. The other part of ABCs is, of course, to ensure that any concrete instance has all the needed methods, etc., so they would ensure that abstract dtypes cannot be instantiated.

seberg replied:
Note to self: It could also be an option to include a version (maybe NPY_API_VERSION maybe another one) in these structs, that way we can add fields at to the end without breaking ABI (with the new fields being safe to ignore on old versions). Note sure I like it, and it would still only allow adding things at the end of the struct.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+* ``is_native(self) -> {0, 1}``
+* ``ensure_native(self) -> dtype``
+* ``default_descr(self) -> dtype`` (return must be native and should normally be a singleton)
+* ``get_sort_function(self, NPY_SORTKIND sort_kind, void **out_sortfunction) -> success {-1, 0}``.
```
Good point :), I am not sold on this part, it was a bit more of an example of the general direction. There was actually the fact that it maybe should return `NotImplemented` if the SORTKIND is not understood.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+* ``common_dtype(cls, other) -> DType, NotImplemented, or NULL``
+* ``common_instance(self, other) -> dtype or NULL``
+
+If not set, most slots are filled with slots which either error or defer automatically.
```
Often yes... But I think the issue is for container types (covariance vs. contravariance). It all confuses me a lot. The main thing is that it is never really *safe* to cast to the super class, but is it always *unsafe*? If I cast from Float64 to UnitFloat64 and it subclasses, should we allow that as "unsafe". Is `nonzero` still correct if I derive an NAInt64 from Int64?

Wikipedia says: Mutable data types which act as both sources and sinks should be invariant (https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)), but to be honest I have not fully grasped the theory. But whenever I tried to think about a good generalization something seemed to make it complicated...

seberg replied:
However, I think it might be fine on AbstractDTypes, and even inheriting "defaults" from the AbstractDType, even for slots/casting! You just must not subclass *concrete* dtypes. I am just thinking we can defer that a bit for now, probably.

mhvk replied:
Hmm, thinking more, I'm now wondering about not being allowed to subclass `float64` - and it may again be good to be super-explicit: I think from what you write as an dtype instance as something that completely describes what some pieces of memory mean - this to me makes sense (though would be good to be explicit how it goes for units like in datetime!). But what about the dtype above? Is it that the class knows what the meaning is, and the instance only defines possible details of the storage? But why would that need to prevent subclassing?

Though probably none of this matters too much - there is real clarity about what the `AbstractDType` is, which I think is the more important distinction.

seberg replied:
I am cautious about subclassing, and still think it is a bad idea. We can make it easy to wrap or fallback to other datatypes, but...
There is actually a *real* and valid use case for subclassing which for sure does not violate Liskov's substitution principle: In theory (not sure how realistic) CuPy, could subclass float64, to add information that they need to work on the GPU, but still share implementation (and even be compatible with) NumPy.
(adding a sentence for that, may get a bit much, but it is an important possibility I think)

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+
+* ``discover_descr_from_pyobject`` (uses ``default_descr`` instead)
+* ``common_instance`` (uses ``default_descr`` instead)
+* ``ensure_native`` (uses ``default_descr`` instead)
```
Ahh! Yes, we could do that. Hmmmmm.

seberg replied:
On the one hand i agree (from the Python API its a nice choice for sure), on the other hand, it does seem a bit clumsy from C maybe? Note that in the end I think I want to store the flag in either case, since, if it is valuable at all, lookup has to be *very* quick in C, and subclass checking is slower (very slow for abstract classes, although I could hand optimize that).

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+This is identical to the proposed API for ufuncs. By default the two dtypes
+are passed in as the last argument. On error return (if no error is set) a
+generic error will be given. Note that right now users can only provide a contiguous
+loop.
```
Yes, my guess would be we allow spec to get things like:
* contiguous_inner_loop
* strided_inner_loop
* scalar_inner_loop

and pick the best one for each case. We could allow users to override the whole thing as well, like a `get_inner_loop` function with a lot additional info.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+
+Although verbose, the API shall mimic the one for creating a new DType.
+The ``PyArrayCastingImpl_Spec`` will include a field for ``dtypes`` and largely
+identical to a ``PyArrayUFuncImpl_Spec``::
```
The ufunc loop will have the same `adjust_dtypes` function. And the UfuncLoop/Impl object may in the future be callable (on ndarray) from python. Which CastingImpl could inherit from it. Further the way the actual inner-loop is fetched, can be identical.

seberg replied:
Hmm, good point, the `casting` is somewhat more central (we need a default one, ufuncs do not need that I think), but really I probably still need a `FromSpec` function, so small differences between the two can be moved there to keep it more similar.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+It should be noted that abstract DTypes are distinct mainly in two ways:
+
+1. They do not have instances. Instantiating an abstract DType has to return
+   a concrete subclass.
```
Yes, you should not be able to. This should be a classmethod such as: `AbstractDType.default_descr()`, which can return an actual dtype. The idea being that `AbstractPyIntegerDType[54].default_descr()` return an int64.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+It allows for the creation of a datatype hierarchy, but avoids issues with
+subclassing concrete DTypes directly:
+Logic such as ``can_cast`` does not cleanly inherit e.g. from a
+``Int64`` to a ``Datetime64`` even though the ``Datetime64`` could be seen
```
Yes, each datetime64 with different unit is its own instance. It is a good example to use throughout, since units are one of the most complex use-cases. I think *unsafe* casting is OK for these (i.e. it is allowed to lose the unit).

mhvk replied:
Great to use it!

Though personally I do not think it should be OK to cast unsafely - you have not just thrown away precision or so, but the actual meaning of the number - just like you would if you cast from meters to seconds - I think this is something you should use  `.view(integer)` for. (In astropy, for any number with units we allow only 0, infinity, and nan to have arbitrary unit - which does break down for 0 for Celsius, but then, that is a bad unit!)

seberg replied:
That is a fair point, maybe we actually need an "even more unsafe" casting :). Note that precision loss in some sense covered by `"same_kind"` casting? I had not thought about it too much that our current "unsafe" casting. However, I think you could also do this:

* `UnitFloat64[m] -> Float64` (cannot cast!)
* `UnitFloat64[unitless] -> Float64` unsafe cast.

Now the question is, how do you get from `UnitFloat64[m]` to `UnitFloat64[unitless]`. We could even decide that you need a UFunc to do that. Or that step is "unsafe" (and a view), and you simply need two steps to get to the actual `float64`. Requiring two steps seems completely fine to me, but it seems like we may still need another level of safeness, basically what you call `view` which currently means: ignore casting safety entirely.

mhvk replied:
Your itemized example is how astropy works! But you are right that of course our quantities do have a `.value` property that just gives the unit-less version (it is just a view), and one uses this quite a bit (as one does the `to_value(new-unit)` which gives the value in the new unit). So, it may have to be that the dtype itself introduces a new casting rule? Maybe those casting strings can become dtype-specific? For now, probably best to leave open!!

seberg replied:
Yes, we could specify (for new slots) that they have to allow for new casting rules to be added in the future. But I actually think that if you are concerned that "unsafe" is not enough, the custom ufunc is not unreasonable: If you make unit, add `units.force_unit(arr, dtype=units.Float64("m"))` ufunc.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+    this flag signals that the data is stored in the default/canonical way.
+    In practice this is always an NBO check, but generalization should be possible.
+    A use case is a complex-conjugated instance of Complex which is a
+    non-canonical representation, but may be native byte order.
```
Nice idea. Canonincal may be a nicer name. I see it as a replacement/generalization `isnative`

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+Coercion to and from Python Objects
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Coercing to and from python scalars requires two functions:
```
Hmmm, the fact that we currently have a bit of a fuzzy concept, seemed that this is maybe improtant enough to have specifically. OTOH, you are right, it is largely a special casting rule (i.e. we could also have these "slots" for convenience, but actually use them to create the `CastingImpl` we actually use).

seberg replied:
I am tending to: lets not worry about it (although if we change our mind on the way, fine). If we assume that most datatypes can keep up with small changes, it is not set in stone in any case.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+
+Coercing to and from python scalars requires two functions:
+
+1. ``__dtype_setitem__(self, item_pointer, value)``
```
It doesn't work on the python side :), at least not for now... I guess it would have to be something like a file-like object as you say..

mhvk replied:
In principle, I guess it could default to `<relevant-piece of memory>.view('u1')` or, equivalently, `array[index].view('u1')`

seberg replied:
Yes, note that I see this not as an actual Python API proposal, but only as a way to reason about it a bit easier then writing the C version.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+   The ``base_obj`` should be ignored normally, it is provided *only* for
+   memory management purposes, pointing to an object owning the data.
+   It exists only to allow support of structured datatypes with subarrays
+   within NumPy, which (currently) return views into the array.
```
Yeah, at least normally it should probably. There is some complexity about how we want user scalars to behave (should they have to be array scalars at all, should we allow them to choose to not use `PyArray_Return`, and if, would reductions return a scalar or not...)
I currently see it: For now they probably should be array scalars (but it likely will be the users responsibility). But we could add at some point that dtypes can indicate that they do not want `PyArray_Return` to be called...

seberg commented:
Thanks a lot for the quick read and feedback, I will try to incorporate it in the next days!

About your general concern: Yes, Units should not specifically inherit from the numerical DTypes (I do not wish to allow inheritance as such). However, they should be able to reuse most of the logic.

I personally envision Unit to be a DType (or rather `UnitFloat64`, `Unit` in my eyes is more like an ABC for all `Unit{Float64, ...}`, although this API would allow both designs), which basically wraps and is able to fallback to Float64 for most logic.
And actually, currently this NEP is still misses that "wrapping" part, because it requires a `PyArray_GetDTypeMetaSlot()` function, to be able to "inherit" (in the sense of a wrapped type), or maybe a `wraps` slot in the `DTypeMetaFromSpec`, which "inherits" some slots (such as sorting, nonzero, ...), but not all (e.g. casting) from the numeric dtype it is based on.

mhvk commented:
A slightly larger question (perhaps really for one of the other NEPs): do the dtypes perhaps conflate two separable things: of how to interpret a given set of bits (float64, int32, etc) and how those are stored in memory (endianness)? If so, might it make sense to explicitly disentangle the two? Perhaps by having separate mixins for that (`DType{Big|Little}Endian`)? Or by having the dtype instance instantiated with its own endian-correcting instance? I think the current scheme is effectively the latter (which seems better), in that code for converting endianness is presumably shared, but perhaps it is good to discuss explicitly.

I think the units are similarly somewhat separate: one can cast from float32[m] to float64[m] or float64[km] but not to any float*[s]. It is not obvious to me if this is best thought of as a mixin unit class or rather a superclass that includes a dtype.

seberg commented:
Well, lets say this:

* My non-flexible DType instances are **only** about how the memory is stored. That is arbitrary, but everything else must represent the exact same thing.
* "Flexible" DTypes are tricky, because yes, they conflate two things: Memory storage, and different meaning or value ranges (units, or string length, etc.).

It would probably be possible to see flexible dtypes as `AbstractDTypes` which are factories for Concrete DTypes instead. The thing about this, though, is that both units and string lengths are an open ended set of instances, and an open ended set of classes/types is much harder to reason about.
All units, or all `Float64[unit]` being a single DType, and you define one `CastingImpl` for it. Of course you *could* do that, you could say that e.g. a`CastingImpl` can be registered on an AbstractDType, but it complicates things a lot. Because the type hierarchy is unclear (if someone subclasses it, can you use the CastingImpl?)

OTOH, yes, you cannot do something like `isinstance(Float64("m"), Length)`...

I never really saw that as a practical choice to be honest, I will think about it a bit. But due to the fact that I wish to be able to memoize `CastingImpl` (and `UfuncImpl`) efficiently, I feel the "flexible" distinction is probably the easier and more efficient concept.

mhvk commented:
@seberg - thanks for the detailed replies! Clearly food for thought. I think for units it will be super-useful to work out the `Datetime` dtype in a bit more detail - as you noted, it is already a quantity, be it based on integer values.

seberg committed:
Add datetime to the dtype hierarchy plot

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+  strings is of course "safe" in general, but may be "same kind" in a specific
+  instance if the second string is shorter. If neither type is flexible this
+  ``adjust_descriptors`` must use it. 
+* ``adjust_descriptors(dtypes_in[2], dtypes_out[2], casting) -> int {0, -1}``
```
Note to self: After talking to Warren, I it may makes sense if we return `casting`, that way we could in theory add new casting more easily. It also pushes the natural place to the construct the casting error out of the `adjust_descriptors` function, where there is more context to construct an error message relevant to the user.

mhvk replied:
Yes, I like that idea.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,796 @@
+  the returned dtype must fullfill ``dtype.is_native() is True`` and be otherwise
+  identical.
+
+* ``DType.type`` is the associated scalar type, also ``dtype.type`` is defined
```
There is a pretty big philosophical and complicated issue around this, which on the one hand I would love to decide and discuss on the mailing list, but on the other hand feel it will be super hard to figure out.

Should scalars be encouraged/discouraged, and do we want array-scalars? For NumPy itself this does not matter, we will not change it. But in principle `dtype.type` could be `None` and in that case we would never return scalars (there would be no scalar after all!). On the other hand, I still somewhat like scalars, so the alternative is that I automatically create a simple scalar type for you if you set type to `None`...

The next thing is: For such new datatypes, we could try to make the ufunc `PyArray_Result` behaviour sane... I would love to just not decide it, but as soon as people create new datatypes there will be an implicit decision.

seberg committed:
Some clarifications and smaller updates (not a proofread and not examples yet)

seberg committed:
clarify how casting will be signalled

seberg committed:
Updates to the NEP, hopefully answer a lot of Martens questions

Also uses datetime64 as an example in a few places, which
is hopefully useful.

rossbar reviewed (commented):
Just some comments from the first half of NEP 42 - I really enjoy the reading, in conjunction with 40 and 41 it's helping me to understand the dtype system in much more detail.

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+
+.. note::
+
+    Python integers do not have a clear/specific NumPy type associated to them
+    right now.  Instead they will need to be be implemented using an
```
```suggestion
    Python integers do not have a clear/specific NumPy type associated with them
    right now.  Instead they will need to be be implemented using an
```
Maybe expand here and explain why (capturing a bit more of the above comment on why Python integers are an interesting example)

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+
+To allow for a DType to accept specific inputs as known scalars, we will
+initially use a method such as ``known_scalar_type`` or ``known_scalar_object``
+which allows discovering a e.g. ``vector`` as a scalar instead of an instance
```
```suggestion
which allows discovering a e.g. ``vector`` as a scalar instead of an instance
```
Instead of "instance" do you mean to say something like "collection of scalars"?

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+For most datatypes, which are not parametric, this is always a canonical default
+instance and ``__discover_descr_from_pyobject__`` does not need to be defined.
+For parametric datatypes, however, such as strings, it is necessary to find the
+correct string length and thus inspect the given python object, this is
```
```suggestion
correct string length and thus inspect the given Python object, this is
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+a dtype instance.
+*Implementors should aim to avoid complex logic in this function.*
+
+After the dtype instance is found, the common dtype with the current previous
+found common dtype instance to find a single dtype that represents all inputs
+and thus support for example mixed integer and floating point input.
```
```suggestion
After the dtype instance is found, the common dtype is compared with the current
previously-found common dtype instance to find a single dtype that represents all
inputs and thus supports, for example, mixed integer and floating point input.
```
Unless I've misunderstood, it might be more clear to say "After the dtype instance is found for each element"

seberg replied:
Added some pseudo-code, maybe that is a thing I generally need to really, too much text, too little pseudo-code, hmmm...

rossbar replied:
If there was one over-arching comment I had from the readthrough, it would be this :). Perhaps the most informative part is the picture of the dtype hierarchy - if some of the text could be replaced with an example with pseudo code I think it could really improve digestibility!

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+
+If one of the objects is an array, or array-like object, its datatype is
+used directly.
+In the case where the user provided a DType, it will be force-cast to the
+user given datatype (class). Force-casting is the current behaviour, and this
+may be made more restrictive in the future.
```
```suggestion
In the case where the user provided a DType, the object will be force-cast to the
user's given datatype (class). Force-casting is the current behaviour, and this
may be made more restrictive in the future.
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+user given datatype (class). Force-casting is the current behaviour, and this
+may be made more restrictive in the future.
+
+In the case that an object is neither a known type, an array-like, or a
+sequence.  The object will be regarded as handled as being of ``object`` dtype.
```
```suggestion
Finally, in the case that an object is neither a known type, an array-like, nor a
sequence, the object will be handled as being of ``object`` dtype.
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+Such a conversion will require passing the correct DType (fixed length of the
+string) or providing a utility function to the user.
+
+The use of a global type map means that an error or warning has to be given
+if two DTypes wish to map to the same python type, in most cases DTypes user
+DTypes should only be implemented for types defined within the same library to
+avoid the potential for conflicts and .
```
```suggestion
The use of a global type map means that an error or warning has to be given
if two DTypes wish to map to the same Python type. In most cases user
DTypes should only be implemented for types defined within the same library to
avoid the potential for conflicts.
```
Removed the "and" at the end but it seems there may have been another thought there you wished to include

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+if two DTypes wish to map to the same python type, in most cases DTypes user
+DTypes should only be implemented for types defined within the same library to
+avoid the potential for conflicts and .
+It is the DType implementors responsibility to be careful about this and use
```
```suggestion
It will be the DType implementor's responsibility to be careful about this and use
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+Alternatively, we could rely on the scalar belonging to the user dtype to
+implement ``scalar.dtype`` or similar.
+Initially, at the moment the exact implementation shall be *undefined*.
+Scalars will have to derive from a NumPy scalar, and provide at least a
+``.dtype`` attribute. Although, this may not be used within NumPy.
```
```suggestion
Scalars will have to derive from a NumPy scalar and provide at least a
``.dtype`` attribute, though, this may not be used within NumPy.
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+would then find correct dtype instance (the second pass is often not necessary).
+The advantage of this is that the DType class information is vital for universal
+functions to decide which loop to execute.
+The first pass would provide the full information necessary for value based
```
```suggestion
The first pass would provide the full information necessary for value-based
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+It seems unlikely that this is useful, and similar to a global, could be
+added later if deemed necessary.
+
+**Further Issues or Discussion:**
```
```suggestion
**Further Issues and Discussion:**
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+
+**Further Issues or Discussion:**
+
+While it is possible to create e.g. a DTypes such as Categorical, array,
```
```suggestion
While it is possible to create e.g. a DType such as Categorical, array,
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+
+    np.array(["2020-01-02", "2020-01-02 11:24"], dtype="M8")
+
+By inspecting the date strings. Together with the below common dtype
+operation this allows it to automatically find that the datetime64 unit
+should be "minutes".
```
```suggestion
By inspecting the date strings. Together with the below common dtype
operation, this allows it to automatically find that the datetime64 unit
should be "minutes".
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+Common DType Operations
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Numpy currently provides functionality of ``np.result_type`` and
+``np.promote_types`` (while more common, ``np.result_type`` has some more
+complex logic due to implementing value based promotion [value_based]_).
+(Note that the name ``np.common_type`` is associated with the scalars and not
+directly with dtypes and is of limited usefulness)
```
```suggestion
Numpy currently provides functions like ``np.result_type`` and
``np.promote_types`` for determining common types [#ct1]_ [#ct2]_

.. [#ct1] While ``np.promote_types` is more common, ``np.result_type`` has some more complex logic due to implementing value-based promotion [value_based]_.
.. [#ct2] Note that the name ``np.common_type`` is associated with the scalars and not directly with dtypes and is of limited usefulness)
```
Since you have two parenthetical statements here, you could try turning them into footnotes instead (I think the syntax I suggested is at least close to how you can do that in rST)

seberg replied:
Just deleted half of it :).

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+here common type operation is used instead of promotion.
+
+**Motivation:**
+This common type operations is vital for the above array coercion when different
+input types are mixed.
```
```suggestion
Common type operations are vital for array coercion when different
input types are mixed.
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+**Motivation:**
+This common type operations is vital for the above array coercion when different
+input types are mixed.
+It also provides the logic currently used to decide the output dtype of
+``np.concatenate()`` and on its own is a useful operation.
```
```suggestion
They also provide the logic currently used to decide the output dtype of
``np.concatenate()`` and on their own are quite useful.
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+It also provides the logic currently used to decide the output dtype of
+``np.concatenate()`` and on its own is a useful operation.
+
+Furthermore, it may be used to find the correct dtype to use for functions with
+different inputs (including universal functions).
```
```suggestion
Furthermore, common type operations may be used to find the correct dtype
to use for functions with different inputs (including universal functions).
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+   (Hypothetical example:
+   ``float_arr + string_arr -> string``, but the output string length is
+   not the same as ``np.concatenate(float_arr, string_arr)).dtype``.)
+2. Array coercion, and concatenate require the common dtype *instance*.   
```
```suggestion
2. Array coercion and concatenation require the common dtype *instance*.   
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+2. Array coercion, and concatenate require the common dtype *instance*.   
+
+**Implementation:**
+The implementation of the common dtype (instance) has some overlap with
+casting.
```
```suggestion
The implementation of the common dtype (instance) determination 
has some overlap with casting.
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+   to `"S32"` (see below for how casting will be defined).
+3. ``__common_instance__("S8", "S32")`` returns the final `"S32"`. 
+
+The main reason for this is to avoid identical functionality in different
```
in different <something missing>

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+which may lead to inconsistent implementations.
+The design (together with casting) naturally separates the concerns of
+different Datatypes.
+Even if tempted, the above Float64 cannot assume it knows how to step 3 correctly.
```
```suggestion
Even if tempted, the above Float64 cannot assume it knows how to resolve step 3 correctly.
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+
+The common type operation cannot be simplified to using only safe casting
+logic.
+As a fall-back testing whether one of the inputs can be safely cast to the
+other could be used when no specific ``__common_dtype__`` is implemented.
+However, this does not allow for the case of::
```
```suggestion
As a fallback, testing whether one of the inputs can be safely cast to the
other could be used when no specific ``__common_dtype__`` is implemented.
However, this does not allow for the case of::
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+
+**Alternatives:**
+
+The use of casting for common dtype (instance) neatly separates the concerns
```
```suggestion
The use of casting for common dtype (instance) determination neatly separates the concerns
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+
+The use of casting for common dtype (instance) neatly separates the concerns
+and allows for a minimal set of duplicate functionality being implemented.
+In cases of mixed DType (classes), it also adds an additional indirection
```
I'm not sure I understand what is meant by "indirection" here

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+The common dtype (of two instances) could thus be implemented explicitly to avoid
+this indirection, potentially only as a fast-path.
+The above suggestion assumes that this is, however, not a speed relevant path,
+since in most cases, e.g. in array coercion, only a single python type (and thus
```
```suggestion
since in most cases, e.g. in array coercion, only a single Python type (and thus
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+There mainly two distinctions:
+
+1. Casting always requires an explicit output datatype to be given.
+2. The NumPy iterator API requires access to lower level functions than
```
```suggestion
2. The NumPy iterator API requires access to lower-level functions than
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+
+1. Casting always requires an explicit output datatype to be given.
+2. The NumPy iterator API requires access to lower level functions than
+   is necessary for current universal functions. 
```
```suggestion
   is currently necessary for universal functions. 
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+
+**Motivation:**
+
+Similar to the common dtype/DType operation above, we again have to use cases:
```
```suggestion
Similar to the common dtype/DType operation above, we again have two use cases:
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+
+DTypes such as a Unit datatype, wrapping an existing Numerical datatype *must*
+be able to access these ``CastingImpl``.
+The will be able to wrap and modify them as necessary. This means that
```
```suggestion
They will be able to wrap and modify them as necessary. This means that
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+If the returned datatypes differ from the provided ones, this means that additional
+casting steps are required for either input or output.
+
+The casting returned by ``adjust_descriptors`` and set on the CastingImpl as
```
```suggestion
The casting returned by ``adjust_descriptors`` and set on the ``CastingImpl`` as
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+The proposed ``CastingImpl`` this designed to be compatible with the
+``UFuncImpl`` proposed in NEP 42.
+While initially it will be a distinct object, the aim is that ``CastingImpl``
+can be a subclass of ``UFuncImpl`` (at least conceptionally).
```
```suggestion
can be a subclass of ``UFuncImpl`` (at least conceptually).
```

rossbar reviewed (commented):
Part 2 of the review I started earlier - again, mostly grammatical and clarification-type suggestions. Thanks for taking the time to put all of this information together!

seberg committed:
Apply suggestions from code review

Commit mainly grammer/style fixes from ross (direct suggestions)

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1149 @@
+
+As a brief note on nomenclature, it should be noted that ``dtype`` normally
+denotes the dtype *instance*, which is the object attached to a numpy array.
+On the other hand the ``DType`` class or type is the subclass of ``np.dtype``.
```
You are right, I should use only `class` here, "type" has a different meaning (and is closer to `dtype.type`.

seberg committed:
Incorporate second chunk of Ross' feedback

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1184 @@
+* ``ensure_canonical(self) -> dtype``
+* ``default_descr(self) -> dtype`` (return must be native and should normally be a singleton)
+* ``get_sort_function(self, NPY_SORTKIND sort_kind) -> {out_sortfunction, NotImplemented, NULL}``.
+  If the sortkind is not understand it may be allowed to return ``NotImplemented``.
```
```suggestion
  If the sortkind is not understood it may be allowed to return ``NotImplemented``.
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1184 @@
+only be extended clumsily (e.g. by adding a version tag to the ``slots``
+to indicate a new, longer version of the struct).
+We could also provide the struct using a function, which however will require
+memory management but would allow ABI compatible extension
```
```suggestion
memory management but would allow ABI-compatible extension
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1184 @@
+This is identical to the proposed API for ufuncs. By default the two dtypes
+are passed in as the last argument. On error return (if no error is set) a
+generic error will be given.
+More optimized loops are in use internally, and made available to users
```
```suggestion
More optimized loops are in use internally, and will be made available to users
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1184 @@
+More optimized loops are in use internally, and made available to users
+in the future (see notes)
+The iterator API currently does not currently support casting errors, this is
+a bug that needs to be fixed. While it is not fixed the loop should always
```
```suggestion
a bug that needs to be fixed. Until it is fixed the loop should always
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1184 @@
+Alternatives
+""""""""""""
+
+Aside from name changes, and possible signature tweaks, there seems few
+alternatives to the above structure.
```
```suggestion
Aside from name changes, and possible signature tweaks, there seem to
be few alternatives to the above structure.
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1184 @@
+pointer inside this structure.
+
+A downside of this is that compilers cannot warn about function
+pointer incompatibilities, there is currently no proposed solution to this.
```
```suggestion
pointer incompatibilities. There is currently no proposed solution to this.
```

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1184 @@
+
+* How ``CastingImpl`` lookup, and thus the decision whether a cast is possible,
+  is defined. (This is speed relevant, although mainly during a transition
+  phase where UFuncs where NEP YY is not yet implemented).
```
A little redundancy in this line

rossbar commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1184 @@
+There is a large space of possible implementations with many discussions
+in various places, as well as initial thoughts and design documents.
+These are listed in the discussion of NEP 40 and not repeated here for
+brievaty.
```
```suggestion
brevity.
```

rossbar reviewed (commented):
Okay - part 3 (and final) comments from reading through the document. Thanks so much for taking the time to put this together. Along with NEP 40 and 41, this document provides a lot of nice insight on dtypes!

seberg committed:
Apply suggestions from code review

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

seberg committed:
Minor adjustment, add example for how promote_types works...

mhvk reviewed (commented):
Had another look, but really too superficially, and I realized some of my comments were answered further down. Perhaps they still help to clarify text...  Probably the only somewhat useful request is for a figure for how casting will work in practice...

mhvk commented:
p.s. The link to the julia docs was very nice!

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1208 @@
+  * DTypes which are not parametric must define a canonical dtype instance
+    which should be a singleton.
+  * Parametric dtypes require some additional methods (below).
+  * One can alternatively create a ``ParametricDType`` class for this purpose,
```
Will look into the image (and other comments).

I think am OK with going there, the question is mainly how to expose it in the C-API I think, for the Python API, I am definitely +1, will add both as options for now. Internally, I think I can assume the flag exists, since I can always set it as part of the metaclass instantiation (similar to python using a flag to implement super-fast subclass checks for some basic types).

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1208 @@
+Another example of methods that should be moved to the DType class are the
+various sorting functions, which shall be implemented by defining a method:
+
+* ``dtype_get_sort_function(self, sortkind="stable") -> sortfunction``
```
Yes, I should note `compare`, which also exists.

We may want to also just delay this, and instead see if we cannot go the gufunc route to eventually allow fast-path sorting for user-dtypes.  The only problem I see with that is that we need a set of internal gufuncs for all different sort-kinds.

mhvk replied:
True, it would definitely need different `gufunc`, but that code already exists, so should be OK. I think @eric-wieser's adaptations of `gufunc` for `np.linalg` are a nice model to follow.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1208 @@
+   within NumPy, which (currently) return views into the array.
+   The function returns an equivalent Python scalar (i.e. typically a NumPy
+   scalar).
+3. ``__dtype_get_pyitem__(self, item_pointer, base_obj) -> object``
```
Well, This is tricky and I may have to expand a bit. In my current branch, I already include casting (for scalars).  This means that `DType.__set_item__()` only needs to correctly be implemented for python types known to the DType (in particular, it must work for its own python scalar-type if that exists).
If the `DType` does not know what to do with the `value`, we will try to find the matching DType and then use casting.

This is currently not really done in NumPy (`float128` has a partial hack), and most types actually can handle everything in theory (since `value.__float__()` is correct for `float16`, `float32` and `float64`).

My current implementation is this:
```
if DType.type is type(value) and  DType.is_known_scalartype(value):
    # By default, should handle all basic python types
    return dtype.__dtype_setitem__(pointer, value)

cast_from_DType = disover_DType_from_pytype(type(value))
if cast_from_DType is None:
     # Not a scalar known to NumPy, so give original setitem a chance.
     return return dtype.__dtype_setitem__(pointer, value)

# We found a DType associated to the scalar type
cast_dtype = DType.discover_descriptor_from_pyobj(value)
tmp_buffer = buffer(size=descr.itemsize)
cast_dtype. __dtype_setitem__(buffer, value)
# cast tmp_buffer and write into pointer.
return
```
will add similar pseudo-code here as well.

mhvk replied:
Probably good to expand as I'm not quite sure where the pseudo-code lives. Also not quite sure about the first `if` statement (why is being the right type not enough and does `value` needs to be checked) or the disappearance of pointer in the last clause. 

But I'll add that I've never even really looked properly at the current implementation, so if my requests for more clarification are becoming too much, do just ignore them!

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1208 @@
+is necessary to provide a low-level C-implementation.
+However, this method shall *not* be part of the public API, instead
+users will initially be limited in what casting functions they can provide
+(e.g. only contiguous loops of multiple items as is used right now).
```
I am currently refusing to do such complex casts. A DType can only define a cast from one of itself to another.

I had currently no intention of "finding a path".  Such direct cast would be either undefined, _or_ one of the DTypes would have to hard-code such a path/fallback.  I would hope that there would be no issue with adding such things later, although I also somewhat hope that it is simply not actually necessary here and would just be over-engineering :).

mhvk replied:
@seberg - definitely fair to keep it simple, and I think as long as there is a general function that does the "search" for a transfer function, it would be extensible anyway.

Still, it may be good to discuss explicitly the straw-man `int24` type that has come up before. If a user were to define it, would there need to be definitions of conversion functions to all numpy `int` and `uint` (and float) types, or just a generic conversion from `int` (and `float`?) and perhaps fast ones from `np.int16` and `np.int8` or so (ie., would it be possible to have conversion if the other `dtype` provided a way to convert to a type in common, say `int` or `float`).

Note that I don't think this actually affects anything for the new `dtype` class, and perhaps it is discussed already in one of the other NEPs... If so, all that's needed here might be a cross-reference.

seberg replied:
Yes, as of my thoughts now `int24` would have to define all of those casts manually.  In theory it could be simple to extend this I guess.
One option could be to allow a CastingImpl (and generally cast) be a multi-step (instead of at most 3 step) process.  You would still have to define a `CastingImpl[int8, int24]`, but internally, it could just point to the same loop as `CastingImpl[int32, int24]`, and its `adjust_descriptor` function would return `int32, int24` instead of `int8,int24`. At which point NumPy would know that it has to do more steps to do the cast.

That might be a bit of a middle ground.  You still need to *define* all the casts, but it can be sufficient to write fewer transfer functions.  If buggy, such a chain of casts may never settle I suppose, but in practice I guess it should be OK and likely not even much more difficult.

I think it is important to note that this is at least a very much possible extension of the API.  The alternative is to have the API getting a full cast-function to actually try to find the graph by itself.  But at that point even the question of whether or not a cast is safe would be a multi-step process.

mhvk replied:
OK, thanks. I indeed think it is extensible, so for now maybe all that is needed is to be explicit about what it is now and that it is possible to extend if the need arises. Clearly, it will never *hurt* to define all the transformations explicitly!

jorisvandenbossche reviewed (commented):
(I am only partly through, but thought to already post my comments, as otherwise they sit here in my browser until I find time to continue. But is was already a good read!)

seberg committed:
Apply suggestions from code review

Apply suggestions from Joris.

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

jorisvandenbossche reviewed (commented):
@seberg the "Casting" section is a bit harder to follow for me (but it's also something I am less familiar with, and more tied to numpy internals, so that's maybe normal). But maybe some more illustrations or pseudocode of a full cast workflow could help.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1208 @@
+**Note:**
+
+The common type operation cannot be simplified to using only safe casting
+logic.
```
I tried to rephrase this a bit. The Casting section (as opposed to just common dtype) below, I rewrote larger chunks... (and added an image, to come soon).

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1208 @@
+
+Additionally, existing methods (and C-side fields) will be provided, although
+fields such as the "kind" and and "char" may be set to invalid values on
+the C-side, and access may error on the Python side.
```
Hmm, very good point. I thought it might error for new dtypes. But you are right, it would be better if it does not.  This means that `"kind"` must return *something* though.

There are two options:

1. The dtype author has to provide a `kind`
2. We simply return the same kind (e.g. None) when it is not defined.

The only problem I see is code which for some reason would do `dtype1.kind == dtype2.kind`, I suppose this is rare, since most will be `dtype.kind == "i"`, etc.  Also on the C-side we are limited to a single struct.  For now I thought we can simply return `\0` for all user dtypes.

In the end, this would mean user dtypes must use `isinstance` checks or similar instead in the long run.

@jorisvandenbossche do you have an opinion on this or does pandas have prior art?

jorisvandenbossche replied:
It's maybe not a problem that it is undefined (or None) for user dtypes, I was mainly thinking about the built-in types like int64. 
In pandas we eg use `kind` as a fast check for certain dtypes (like `dtype.kind in ("u", "i")` for integer dtypes). 

For our own dtypes, we define a "kind" on the dtype object (for compatibility I assume), but it is not always clear what this should be or how it is useful. I think we mostly took the the numpy one if there is a direct equivalent (eg our masked integer dtype also gives kind="i"/"u"), or otherwise just "O" for things like categorical and interval dtypes.

mhvk replied:
Maybe the default should be to return the `dtype` itself (or perhaps its name!?), so it is guaranteed to compare equal only to itself?

seberg committed:
Incorporate feedback from Marten and Joris

Mainly adds a figure for the casting section, and tries to rephrase
the first part of it almost entirely around that figure, which should
also simplify it a lot, although at this time possible loses a tiny
bit precision/details.

seberg commented:
Just a heads up, I think I will merge this tomorrow or so after some updates from Ross and minor fix-ups with the `.kind` as well as having a quick look over myself.  I hate to "lose" part of the discussion, but on the other hand I think its not in a terrible state and the rendered version will be nicer.  Plus the many comments are becoming a bit unwieldy.

seberg commented:
OTOH, merging loses the ability to in-line comment easily, so maybe nvm for now...

mattip commented:
Merging would allow more people to see the progress including the images.

rossbar committed:
NEP: Apply suggestions from Ross

* SUG: Update wording in intro note.

* TYPO: add missing preposition

* TYPO: fix a plural

* SUG: calling the dtype constructor \*returns\* a dtype?

* SUG: terminology types->dtypes (not sure)

* SUG: perhaps a more precise wording?

* TYPO: removed extra comma

* FMT: switch to \*\* syntax for bold.

For whatever reason, _word_ was not creating the proper emphasis
in the html

* TYPO

* SUG: Wording suggestion on inheriting from concrete DTypes

* SUG: punctuation suggestions for DTypeMeta class sentence

* TYPO: word ordering

* TYPO: redundant "instead"

* SUG: Word ordering incorrect?

* TYPO: use-case and an->a

* TYPO: usecase -> use-case

* SUG: emphasize that youre referring to the class attribute

* TYPO: add comma

* TYPO: hyphenate some compound adjectives

* TYPO: fix plural

* SUG: No And to start sentence

* TYPO: Added missing the

* TYPO: Add article to vector dtype

* SUG: remove back-referencing

* TYPO: a -> an

* TYPO: Fix missing dunder

* TYPO: added missing verb

* TYPO: possibly -> a possibility

* FMT: Missing ` for sphinx ref?

Probably still won't work without a caption defined or numref
enabled

* TYPO: and -> an

* FMT: Add missing ` for inline literal

* FMT: Another sphinx link suggestion, but might need a role

* SUG: Wording

* TYPO: hyphenate an adjective

* SUG: Not quite sure if this is correct?

* TYPO: this -> the

* TYPO: stirded->strided

* TYPO: too many currentlys

* TYPO: comma to colo

eric-wieser commented:
Can we add a circle-ci build for NEPs?

seberg committed:
NEP: Fix the last section and missing link.

Also clarify the ``kind`` and ``char`` a bit, suggesting to
use ``__qualname__``. While this is weird for ``char``, it
is at least guaranteed to be be unique and to fail when used
inapproprioatly.

eric-wieser commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1323 @@
+To avoid duplication, but also expose the classes conveniently to users
+we propose the addition of::
+
+    np.dtype[np.int64]
+
+as a class getter. This can work both for user and NumPy DTypes,
+although, in many cases a library may choose to provide a more direct
+way to access the specific DType class.
+This method may initially be limited to concrete DTypes. 
+The main reason for this choice is to provide a single
+clear and future-proof way to find the DType class given the
+Python (scalar) class.
```
Are we sure we want to introduce a second way to spell this, given that `np.dtype(cls)` is already a well-established mechanism to convert a scalar type to a dtype?

seberg replied:
No, I am not sure about this, it is more of a thought and I could reduce or even remove this for now.

There is a generalization whether we should after all consider to create `parametric` instances (similar to our "S0").
I had decided against this, but passing around parametric instances has its advantages (since for some functions the input can be a `dtype` instance _or_ a `DType` class).
But on the other hand, means that the neat split between: instances can always be attached arrays, classes cannot. Is gone.

Note that there are some subtleties around this. E.g. `np.dtype(np.float64)` enforces that you do get native byte-order (the default instances), while `np.dtype[np.float64]` (or however you spell it if you want to) will *not* enforce that if the input is already a `>float64`.

So the question is whether `np.dtype(Unit)` should be allowed and return something which cannot be attached to an array...

I guess the main issue is about `issubdtype` checks in the long run.  E.g. how to check whether a dtype is `inexact`, since `isinstance(dtype, np.dtype[np.inexact])` is not very cute (of course the `dtype.type` way/subdtype stays valid.

eric-wieser commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1323 @@
+   It may return ``NotImplemented`` to defer to ``other``.
+   (For abstract DTypes, subclasses get precedence, concrete types are always
+   leaves, so always get preference or are tried from left to right). 
+2. ``__common_instance__(self, other : cls) -> cls`` is used when two instances
```
What is `cls` here? Is it the type of `self` too? Perhaps

```suggestion
2. ``__common_instance__(self: SelfT, other : SelfT) -> SelfT`` is used when two instances
```

eric-wieser commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1323 @@
+The solution is partially "outsourced" to the casting machinery by
+splitting the operation up into three steps:
+
+1. ``__common_dtype__(type(>float64), type(S8))`` returns `String`.
+2. The casting machinery provides the information that `">float64"` casts
+   to `"S32"` (see below for how casting will be defined).
+3. ``__common_instance__("S8", "S32")`` returns the final `"S32"`. 
```
can you specify what objects these methods are called upon?

eric-wieser reviewed (commented):
Haven't read through the whole thing of many of the existing comments, but thought I'd at least make a start.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1323 @@
+2. Abstract DTypes can enable code such as
+   ``arr.astype(Complex)`` to express the desire to cast to a
+   complex data type of any precision.
+3. It anticipates the creation of families of DTypes by users.
+   For example allowing the creation of an abstract ``Unit`` class with a concrete
+   ``Float64Unit``. In which case ``Unit(np.float64, "m")`` could be
+   identical to ``Float64Unit("m")``.
```
Well, I think its open for the implementer, and I agree that `Unit(np.float64, "m")` should give a concrete `dtype` instance (I assume this is what you mean).  I personally think the existence of a `Float64Unit` as a DType class may make sense, because it seems to me it would make dispatching simpler (and a single step).  However, I agree that this is not strictly necessary. I will try to shorten the example a bit.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1323 @@
+A very concrete example is the current Pandas ``Categorical`` DType
+which may benefit from abstraction to allow the differentiation of
+a categorical of integer values and one of general object values.
+The reason for this is that we may want to reject
+``common_dtype(CategoricalInt64, String)`` while
+``common_dtype(CategoricalObject, String)`` is valid and the result
+will have the ``object`` DType.
+While the current ``Categorical`` combines both and must still be represented.
```
Thanks, I had just fixed the last sentence earlier, used your wording its much nicer.

seberg committed:
NEP: Type edits for NEP 42 and Eric Wiesers suggestions

Some type edits for the first part mainly and fixups based on
Eric's review.

seberg commented on doc/neps/nep-0042-new-dtypes.rst:
```diff
@@ -0,0 +1,1323 @@
+    stores ``real - imag`` instead of ``real + imag`` and is thus not
+    the canonical storage.
+
+* ``ensure_canonical(self) -> dtype`` return a new dtype (or ``self``).
```
Ah, I called it ensure, because we currently have an `EnsureNBO` function or so. I did not update it yet, but definitely an option.

seberg committed:
Fixup links

seberg commented:
OK, this PR is getting very unwieldy (at least in my browser) and I got at least some positive feedback on it, so I will merge it now.  I hope everyone will be fine with that (Sorry Eric if you intended to get the slightly easier back/and forth in here still), but we can still do at least one more pass as inline comments in this PR!

The final link (as soon as it is published, which should be very quick) will be:

https://numpy.org/neps/nep-0042-new-dtypes.html

It is not like merging finalizes this in any serious way yet, I am open to do larger modifications!

I realized I never edited the metadata... I think especially @mhvk, but also  @jorisvandenbossche, @eric-wieser, and @rossbar all helped a lot, and I think you should be part of the author lists unless you do not want to be.

mhvk commented:
Makes sense to get it in. Am happy to be listed as contributor, though all kudos really to you for sticking with us! It is very tricky and as has been said by multiple commenters, the NEP gives a good sense of why it is tricky and how to think about the dtypes.

jorisvandenbossche commented:
> merging loses the ability to in-line comment easily, so maybe nvm for now...

BTW, you can normally "reopen" a merged PR, in case some want to keep commenting inline
